apps/op-app/src/adapters/io/user-metadata.ts

```diff
@@ -3,49 +3,126 @@ import {
   UserMetadata,
   userMetadataSchema,
 } from "@/domain/user-metadata.js";
+import { NonEmptyString } from "@pagopa/ts-commons/lib/strings.js";
 import * as E from "fp-ts/lib/Either.js";
 import * as assert from "node:assert/strict";
-import { ZodError } from "zod";
+import { ZodError, z } from "zod";
 
-import { Client, createClient } from "./generated/client.js";
+import { IOConfig } from "./config.js";
+import { AssertionRef } from "./generated/lollipop/AssertionRef.js";
+import { LollipopAuthBearer } from "./generated/lollipop/LollipopAuthBearer.js";
+import * as Lollipop from "./generated/lollipop/client.js";
+import * as SessionManager from "./generated/session-manager/client.js";
+
+export class UserNotLoggedError extends Error {
+  name = "UserNotLoggedError";
+  constructor() {
+    super("User not logged");
+  }
+}
 
 export class IO implements IdentityProvider {
-  #client: Client;
+  #lollipop: Lollipop.Client;
+  #lollipopApiKeyAuth: string;
+  #sessionManager: SessionManager.Client;
 
-  constructor(baseUrl: string) {
-    this.#client = createClient({
-      baseUrl,
-      fetchApi: globalThis.fetch,
+  constructor(config: IOConfig) {
+    const fetchApi = globalThis.fetch;
+    this.#sessionManager = SessionManager.createClient({
+      baseUrl: config.sessionManager.baseUrl,
+      fetchApi,
     });
+    this.#lollipop = Lollipop.createClient({
+      baseUrl: config.lollipop.baseUrl,
+      fetchApi,
+    });
+    this.#lollipopApiKeyAuth = config.lollipop.apiKey;
   }
 
-  async getUserMetadata(token: string): Promise<UserMetadata> {
+  async #getAssertion(
+    lollipopAuth: LollipopAuthBearer,
+    assertionRef: AssertionRef,
+  ): Promise<string> {
     try {
-      const result = await this.#client.getUserForFIMS({
-        Bearer: `Bearer ${token}`,
+      const response = await this.#lollipop.getAssertion({
+        ApiKeyAuth: this.#lollipopApiKeyAuth,
+        assertion_ref: assertionRef,
+        "x-pagopa-lollipop-auth": lollipopAuth,
       });
-      assert.ok(E.isRight(result));
-      assert.strictEqual(result.right.status, 200);
-      const {
-        family_name: lastName,
-        fiscal_code: fiscalCode,
-        name: firstName,
-      } = result.right.value;
-      return userMetadataSchema.parse({ firstName, fiscalCode, lastName });
-    } catch (e) {
-      const err = {
-        cause: "Client error",
-        msg: "Unable to fetch user data",
-      };
-      if (e instanceof ZodError) {
-        err.cause = "Invalid user data";
-      } else if (e instanceof assert.AssertionError) {
-        err.cause = "Request failed";
-      }
-      const msg = `${err.msg}. ${err.cause}.`;
-      throw new Error(msg, {
-        cause: err.cause,
+      assert.ok(E.isRight(response));
+      assert.strictEqual(response.right.status, 200);
+      const assertionSchema = z.object({
+        response_xml: z.string().min(1),
+      });
```

Comment:
Can't we extract this to avoid redeclaring this at every `getAssertion`?

Reply:
where is redeclared?

Reply:
I mean that this `const` will be redeclared every time `getAssertion` is called, since the schema is equal for every call we could declare it outside of the method